src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -527,6 +556,38 @@ private static string GetIdentityIfDefaultCredentialsUsed(bool defaultCredential
             return defaultCredentialsUsed ? CurrentUserIdentityProvider.GetIdentity() : string.Empty;
         }
 
+        public static int GetMaxHttp11ConnectionsPerPool()
+        {
+            int max = 0;
+            foreach ((WeakReference<HttpConnectionPoolManager> weakConnectionPoolManagerRef, _) in s_allConnectionPools)
+            {
+                if (weakConnectionPoolManagerRef.TryGetTarget(out HttpConnectionPoolManager? connectionPoolManager))
+                {
+                    foreach ((_, HttpConnectionPool connectionPool) in connectionPoolManager._pools)
+                    {
+                        max = Math.Max(max, connectionPool.Http11ConnectionCount);
```

Comment:
I'd suggest to consider using an associative [MaxHeap](https://en.wikipedia.org/wiki/Binary_heap) (i.e. MaxHeap combined with Dictionary to access elements for O(1)) where each node stores open connection count for a single HttpConnectionPool.

On connection open or close, the pool will increase or decrease the respective counter. Since each pool will change only their own counter, it can be easily done thread safely. When the polling counter `_maxHttp11ConnectionsPerPoolCounter ` triggers, it will call Heapify method to rebuild the heap and get the max counter on the top. Heapify runs in O(logn) in the worst case so it seems fast enough.

There is one more issue though, that is to ensure Heapify can run concurrently with heap nodes changes. To solve that, I'd suggest to guard Heapify call with `lock`, but allow concurrent changes to counters on heap as explained above.

Reply:
Thank you for the idea! While having a better O() when collecting the maximum on polling intervals, I am doubtful it would be a measurable perf improvement at the end, especially considering the added complexity.

The current approach tries to minimize the overhead when telemetry is disabled. The proposed one would incur an overhead on every connection even when telemetry is disabled.

While the O(N) for this collection might seem scary, it only runs once per polling interval, and the N is effectively capped at the port limit, so I doubt it would show up as a significant overhead.

Reply:
> The proposed one would incur an overhead on every connection even when telemetry is disabled.

That's not exactly so. If telemetry is disabled increase/decrease counter methods will be no-op and Heapify will never be called since timer will be disabled as well.

However, I do agree it adds more complexity.